pilot/pkg/status/reporter.go

```diff
@@ -0,0 +1,226 @@
+		Reporter:            r.PodName,
+		DataPlaneCount:      len(r.status),
+		InProgressResources: map[string]int{},
+	}
+	// for every version (nonce) of the config currently in play
+	for nonce, dataplanes := range r.reverseStatus {
+		// for every resource in flight
+		for _, res := range r.inProgressResources {
+			// check to see if this version of the config contains this version of the resource
+			// it might be more optimal to provide for a full dump of the config at a certain version?
+			key := res.String()
+			if dpVersion, err := r.store.GetResourceAtVersion(nonce,
+				res.ToModelKey()); err == nil && dpVersion == res.ResourceVersion {
+				if _, ok := out.InProgressResources[key]; !ok {
+					out.InProgressResources[key] = len(dataplanes)
+				} else {
+					out.InProgressResources[key] += len(dataplanes)
+				}
+			} else if err != nil {
+				scope.Errorf("Encountered error retrieving version %s of key %s from Store: %v", nonce, key, err)
+			}
+			if out.InProgressResources[key] >= out.DataPlaneCount {
+				// if this resource is done reconciling, let's not worry about it anymore
+				finishedResources = append(finishedResources, res)
+				// deleting it here doesn't work because we have a read lock and are inside an iterator.
+				// TODO: this will leak when a resource never reaches 100% before it is replaced.
+				// TODO: do deletes propagate through this thing?
+			}
+		}
+	}
+	return out, finishedResources
+}
+
+// For efficiency, we don't want to be checking on resources that have already reached 100% distribution.
+// When this happens, we remove them from our watch list.
+func (r *Reporter) removeCompletedResource(s []Resource) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	var result []Resource
+	hashtable := make(map[Resource]bool)
+	for _, item := range s {
+		hashtable[item] = true
+	}
+	for _, ipr := range r.inProgressResources {
+		if _, ok := hashtable[ipr]; !ok {
+			result = append(result, ipr)
+		}
+	}
+	r.inProgressResources = result
+}
+
+// This function must be called every time a resource change is detected by pilot.  This allows us to lookup
+// only the resources we expect to be in flight, not the ones that have already distributed
+func (r *Reporter) AddInProgressResource(res model.Config) {
+	myRes := ResourceFromModelConfig(res)
+	if myRes == nil {
+		scope.Errorf("Unable to locate schema for %v, will not update status.", res)
+		return
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.inProgressResources = append(r.inProgressResources, *myRes)
+}
+
+// generate a distribution report and write it to a ConfigMap for the leader to read.
+func (r *Reporter) writeReport(ctx context.Context) {
+	report, finishedResources := r.buildReport()
+	go r.removeCompletedResource(finishedResources)
+	//write to kubernetes here.
+	reportbytes, err := yaml.Marshal(report)
+	if err != nil {
+		scope.Errorf("Error serializing Distribution Report: %v", err)
+		return
+	}
+	r.cm.Data["keyfield"] = string(reportbytes)
```

Comment:
this will panic the first write

Reply:
cm is initialized at line 70.  is there some other reason this would panic?